lib/reversetunnel/localsite.go

```diff
@@ -281,38 +281,45 @@ func (s *localSite) getConn(params DialParams) (conn net.Conn, useTunnel bool, e
 	if params.To != nil {
 		dreq.Address = params.To.String()
 	}
+
 	// If server ID matches a node that has self registered itself over the tunnel,
-	// return a connection to that node. Otherwise net.Dial to the target host.
-	conn, err = s.dialTunnel(dreq)
-	if err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, false, trace.Wrap(err)
-		}
+	// return a tunnel connection to that node. Otherwise net.Dial to the target host.
+	conn, tunnelErr := s.dialTunnel(dreq)
+	if tunnelErr == nil {
+		return conn, true, nil
+	}
 
-		// Connections to applications should never occur over a direct dial, return right away.
-		if params.ConnType == services.AppTunnel {
-			return nil, false, trace.ConnectionProblem(err, "failed to connect to application")
-		}
+	if !trace.IsNotFound(tunnelErr) {
+		return nil, false, trace.Wrap(tunnelErr)
+	}
 
-		// This node can only be reached over a tunnel, don't attempt to dial
-		// remotely.
-		if params.To.String() == "" {
-			return nil, false, trace.ConnectionProblem(err, "node is offline, please try again later")
-		}
-		tunnelErr := trace.Errorf("dialing through a tunnel: %v", err)
-		// If no tunnel connection was found, dial to the target host.
-		dialer := proxy.DialerFromEnvironment(params.To.String())
-		conn, err = dialer.DialTimeout(params.To.Network(), params.To.String(), defaults.DefaultDialTimeout)
-		if err != nil {
-			return nil, false, trace.NewAggregate(tunnelErr, trace.Errorf("dialing directly: %v", err))
-		}
+	s.log.WithError(tunnelErr).Debug("Error occurred while dialing through a tunnel.")
+
+	// Connections to applications should never occur over a direct dial, return right away.
+	if params.ConnType == services.AppTunnel {
+		return nil, false, trace.ConnectionProblem(tunnelErr, "failed to connect to application")
+	}
+
+	offlineMsg := "the node is offline or has disconnected, if the issue " +
+		"persists, restart the node or re-register it in the cluster"
+
+	// This node can only be reached over a tunnel, don't attempt to dial
+	// remotely.
+	if params.To.String() == "" {
+		return nil, false, trace.ConnectionProblem(tunnelErr, offlineMsg)
+	}
 
-		// Return a direct dialed connection.
-		return conn, false, nil
+	// If no tunnel connection was found, dial to the target host.
+	dialer := proxy.DialerFromEnvironment(params.To.String())
+	conn, directErr := dialer.DialTimeout(params.To.Network(), params.To.String(), defaults.DefaultDialTimeout)
+	if directErr != nil {
+		s.log.WithError(directErr).Debug("Error occurred while dialing directly.")
+		aggregateErr := trace.NewAggregate(tunnelErr, directErr)
+		return nil, false, trace.ConnectionProblem(aggregateErr, offlineMsg)
```

Comment:
If the tunnel wasn't found and we fail to dial directly as well, will we still get an ugly error? Maybe return a better worded-error here as well.

Reply:
No, we should get the new message in both cases.